### ISSUE TITLE
Async issue potentially leading to memory leaks.

### ISSUE THREAD

**Issue by sskodje** [XamlAnimatedGif/XamlAnimatedGif#23]
Hello!

When the Animator class is instantiated in AnimationBehavior, it is done in an async way.
```
                var animator = await Animator.CreateAsync(image, sourceUri, repeatBehavior);
```
where it fetches the image from wherever the URI is pointed.

The issue i found, was that one might load an URI as animatedsource on an image. This might take a few milliseconds of an image in memory, or it could take several seconds over the internet. Before the source is resolved and an Animator class is created, you might have closed/unloaded the image again and disposed of the animated source to free up the resources.

If you happen to unload the image before the Animator is created asynchronously by  passing null to SourceUri (i assume this is the intended way to do it?):

```
        private static void ClearAnimatorCore(Image image)
        {
            var animator = GetAnimator(image);
 //this will be null, because it is still querying the image source
// from wherever and hence haven't created the animator instance yet
//, and it returns
            if (animator == null)
                return;

            animator.Dispose();
            SetAnimator(image, null);
        }
```

Now after the failed disposing of the Animator , the animator decides to finish and returns, and does 
```
SetAnimatorCore(image, animator);
```
I think the problem here is twofold. First off, the animator instance is now attaching to an unloaded image, that is not garbage collected yet(?). Secondly, it is passing a reference to the Image into the Animator class, preventing garbage collection of the unloaded image. This image + Animator would live as long as the parent application does.

I'm really not great at object lifetimes etc and i could be wrong, but i think something like this in InitAnimationAsync could work:

```
                var animator = await Animator.CreateAsync(image, sourceUri, id, repeatBehavior);
                if (!image.IsLoaded)
                    throw new InvalidOperationException("Image is not loaded, can't set animator core");
``` 
To prevent the Animator attaching to unloaded Images.

**Comment by thomaslevesque** [comment]
Hi Sverre,


I did notice a leak a few days ago, when changing the current image very quickly in the demo app, but didn’t have time to investigate where it was coming from. Your analysis seems correct. I’ll look into it as soon as I can, but I’m on vacation for a few days with no access to my dev PC, so it will have to wait until next week.


Thanks for your bug reports by the way. You found 3 serious bugs in less than a week, that’s pretty good 😉






From: Sverre Skodje
Sent: ‎Thursday‎, ‎August‎ ‎20‎, ‎2015 ‎4‎:‎25‎ ‎PM
To: XamlAnimatedGif/XamlAnimatedGif





Hello!

When the Animator class is instantiated in AnimationBehavior, it is done in an async way.
                var animator = await Animator.CreateAsync(image, sourceUri, repeatBehavior);


where it fetches the image from wherever the URI is pointed.

The issue i found, was that one might load an URI as animatedsource on an image. This might take a few milliseconds of an image in memory, or it could take several seconds over the internet. Before the source is resolved and an Animator class is created, you might have closed/unloaded the image again and disposed of the animated source to free up the resources.

If you happen to unload the image before the Animator is created asynchronously by passing null to SourceUri (i assume this is the intended way to do it?):
        private static void ClearAnimatorCore(Image image)
        {
            var animator = GetAnimator(image);
 //this will be null, because it is still querying the image source from wherever and hence haven't created the animator instance yet, ans so it returns
            if (animator == null)
                return;

            animator.Dispose();
            SetAnimator(image, null);
        }


Now after the failed disposing of the animator core, the animator decides to finish and returns, and does 
SetAnimatorCore(image, animator);


I think the problem here is twofold. First off, the animator instance is now attaching to an unloaded image, that is not garbage collected yet. Secondly, it is passing a reference to the Image into the Animator class, preventing garbage collection of the unloaded image. This image + Animator would live as long as the parent application does.

I'm really not great at object lifetimes etc and i could be wrong, but i think something like this in InitAnimationAsync could work:
                var animator = await Animator.CreateAsync(image, sourceUri, id, repeatBehavior);
                if (!image.IsLoaded)
                    throw new InvalidOperationException("Image is not loaded, can't set animator core");


To prevent the Animator attaching to unloaded Images.

**Comment by sskodje** [comment]
Hehe, i'm working on migrating my application from WpfAnimatedGif to XamlAnimatedGif, and just bumped into these little problems on the way. I guess the app was a good stress tester :)

Got rid if memory usage into the hundreds though..

**Comment by thomaslevesque** [comment]
OK, just had a look at this issue. Actually I don't think this is related to the Image control being unloaded, because I can reproduce it in the demo app, where the image control is never unloaded.

I see a possible reentrancy issue that could cause an Animator to not be disposed when it should, but it should still be garbage collected (which would cause the finalizer to call Dispose). The only place where the animator is still referenced is in an async state machine, which means there must be an async method still running where the animator is used. But I still don't understand why this method is running... I need to do a more thorough memory profiling to understand (so far I only looked at the VS2015 diagnostic tools, which are nice but not as complete as ANTS or dotTrace...)

**Comment by thomaslevesque** [comment]
Hi Sverre,

It should be fixed in 1.0.3; I'm no longer able to reproduce the leak. The fix is to check if the source has changed after CreateAsync returns; if it has, just dispose the newly created `Animator` object.
